### ISSUE TITLE
Specifying both `width` and `height` incorrectly applies `rect`

### ISSUE THREAD

**Issue by coreyward** [sanity-io/image-url#32]
This library incorrectly sets the `rect` parameter when both `width` and `height` are specified. This prevents many `fit` modes from working properly. [An in-depth visual comparison and breakdown of this bug is available in this CodeSandbox project](https://codesandbox.io/s/sanity-image-url-playground-iioc2). At a glance, this screenshot from the linked breakdown demonstrates the issue:

<img width="752" alt="Screen Shot 2021-02-12 at 4 25 13 PM" src="https://user-images.githubusercontent.com/81224/107828635-e31ae780-6d4e-11eb-9e25-63e76bca62ea.png">

I considered stripping the `rect` prop from the returned urls, but it still needs to be present in two scenarios:

1. **When a source has an associated crop**
  The rationale here should be self-evident: `rect` is used to apply crop values.
2. **When the `fit` mode is `min` and the source has a focal point** 
  This library transforms `hotspot` values into focal point parameters `fp-x` and `fp-y`, but due to a bug/oversight in the Images API, these parameters are only used when the `fit` mode is set to `crop`. Since the `min` mode is identical to `crop` except that the image will not be scaled up, the focal point needs to be emulated via `rect` (or the bug in the Images API needs to be fixed)

I believe this library should only add a `rect` parameter in these two scenarios, and to try to pre-compute and apply the focal point in the second case.

**Comment by johnnywalker** [comment]
Since the API is [modeled after imgix's API](https://sanity-io-land.slack.com/archives/C9Z7RC3V1/p1597143527469900?thread_ts=1597135551.460500&cid=C9Z7RC3V1), I believe `rect` is calculated incorrectly when determined from a source with crop dimensions. Sanity's API appears to treat `rect` as the source rectangle, and [urlForImage.ts](https://github.com/sanity-io/image-url/blob/master/src/urlForImage.ts#L130) constrains the source's crop to the final aspect ratio instead. 

Reference: https://docs.imgix.com/apis/rendering/size/rect

**Comment by modulareverything** [comment]
Sharing this here as it solved the problem for me:

https://github.com/sanity-io/image-url/issues/48#issuecomment-1351987836

In summary, try adding `ignoreImageParams()` to your image URL like so:

`imageUrlBuilder.image(source).ignoreImageParams().width(300).height(50).fit('clip').url()`

**Comment by coreyward** [comment]
@mathiasvalentin-thg Not sure what problem you have, but if you check the playground/demo I built in the original post you can see that the problematic URLs already have `w` and `h` params. The issue is that they also have a `rect` param. It's easy enough to remove the `rect` param and value if you need an interim solution, but this remains errant behavior that produces incorrect outputs per the spec.

**Comment by CunningDJ** [comment]
Running into this same problem.  We have an image well below 2000px in both dimensions - 525x279 -  and use the `imageUrlBuilder.image(xxx).width(2000).height(2000)` with no `rect`, and this triggers a 1:1 ratio and that triggers a rect crop of 279x279, shaving off the left and right side.  Then, the final size we get is 1080x1080 (the pixel limit when Sanity processes based on the largest dimension of an image?)

Is this is intended behavior?  I think part of the confusion might be whether `width()` and `height()` are stating some kind of max width and height, and how they interact with each other.  In our case, I think we can take both modifiers out since I believe they were put in there to bring down image size, and Vercel/NextJS is doing that anyway.

**Comment by coreyward** [comment]
@CunningDJ You still need both `width` and `height` if you want the image in a different aspect ratio (e.g., if you want to crop the image in some form or fashion from its source dimensions to fit into a specific box in the layout). 

I recently released a new package called [`@sanity-image/url-builder`](https://www.npmjs.com/package/@sanity-image/url-builder) to generate either a `src` or `srcSet` from a Sanity image that completely sidesteps this bug. You can use it independently of any other library, or in conjunction with Next Image if desired. It's already baked into `sanity-image` as well if you are using React.

**Comment by CunningDJ** [comment]
I was taking the perspective that for their automated URL builder logic that I read through, it looks like they look at width and height and decide that this equates to a stated crop ratio.  However, it sounds like your point is that unless you explicitly state `rect()`, they shouldn't be adding it.  Without looking at all of the branches inside that function.  I think that's a fair statement.  Have you thought about submitting a PR [somewhere around here](https://github.com/sanity-io/image-url/blob/main/src/urlForImage.ts#L151) instead of adding a separate library?

**Comment by coreyward** [comment]
Yeah, I started work on a PR around the time I opened this issue (4 years ago), but it quickly wound up requiring internal documentation and decision making on how they wanted various things to behave. The `@sanity-image/url-builder` was actually built as part of `sanity-image`, and extracted into a separate library more recently due to many requests for it on the Sanity.io community Slack. Besides, as you mention, changing this behavior would be a breaking change and merit a major version bump, which is likely why it's not been addressed yet.
